enterprise/frontend/src/metabase-enterprise/content_translation/utils.ts

```diff
@@ -0,0 +1,157 @@
+) => string | MsgidType;
+
+/** Translate a user-generated string
+ *
+ * Terminology: A "msgid" is a 'raw', untranslated string. A "msgstr" is a
+ * translation of a msgid.
+ * */
+export const translateContentString: TranslateContentStringFunction = (
+  dictionary,
+  locale,
+  msgid,
+) => {
+  if (!locale) {
+    return msgid;
+  }
+
+  if (typeof msgid !== "string") {
+    return msgid;
+  }
+
+  if (!msgid.trim()) {
+    return msgid;
+  }
+
+  const msgstr = dictionary?.find(
+    (row) => row.locale === locale && row.msgid === msgid,
+  )?.msgstr;
+
+  if (!msgstr || !msgstr.trim()) {
+    return msgid;
+  }
+
+  return msgstr;
+};
+
+const isRecord = (obj: unknown): obj is Record<string, unknown> =>
+  _.isObject(obj) && Object.keys(obj).every((key) => typeof key === "string");
+
+/** Walk through obj and translate any display name fields */
+export const translateDisplayNames = <T>(
+  obj: T,
+  tc: ContentTranslationFunction,
+  fieldsToTranslate = ["display_name", "displayName"],
+): T => {
+  const traverse = (o: T): T => {
+    if (Array.isArray(o)) {
+      return o.map((item) => traverse(item)) as T;
+    }
+    if (isRecord(o)) {
+      return Object.entries(o).reduce((acc, [key, value]) => {
+        const newValue =
+          fieldsToTranslate.includes(key as string) && typeof value === "string"
+            ? tc(value)
+            : traverse(value as T);
+        return I.assoc(acc, key, newValue);
+      }, o);
+    }
+    return o;
+  };
+  return traverse(obj);
+};
+
+export const shouldTranslateFieldValuesOfColumn = (col: DatasetColumn) =>
+  ["type/Category", "type/Country", "type/State", "type/Source"].includes(
+    col?.semantic_type || "",
+  );
+
+export const translateFieldValuesInHoveredObject = (
+  obj: HoveredObject | null,
+  tc?: ContentTranslationFunction,
+) => {
+  if (!tc) {
+    return obj;
+  }
+  return {
+    ...obj,
+    data: obj?.data?.map((row) => {
+      const { value, col } = row;
+
+      return {
+        ...row,
+        value:
+          col &&
+          shouldTranslateFieldValuesOfColumn(col) &&
+          typeof value === "string"
+            ? tc(value)
+            : value,
+      };
+    }),
+  };
+};
+
+export const translateFieldValuesInSeries = (
+  series: Series,
+  tc: ContentTranslationFunction,
+) => {
+  return series.map((singleSeries) => {
+    if (!singleSeries.data) {
+      return singleSeries;
+    }
+    const { rows, cols } = singleSeries.data;
+    const indexesOfColsToTranslate = cols.reduce<number[]>(
+      (acc, col, index) =>
+        shouldTranslateFieldValuesOfColumn(col) ? [...acc, index] : acc,
+      [],
+    );
+
+    const translatedRows = rows.map((row) =>
+      row.map((value, index) =>
+        indexesOfColsToTranslate.includes(index) && typeof value === "string"
+          ? tc(value)
+          : value,
+      ),
+    );
+    return {
+      ...singleSeries,
+      data: { ...singleSeries.data, rows: translatedRows },
+    };
+  });
+};
```

Comment:
For performance it might be sensible to early-exit this function if there are no translations